### ISSUE TITLE
SyncPlay desync when transcoding

### ISSUE THREAD

**Issue by Reinachan** [jellyfin/jellyfin-web#6210]
**Describe The Bug**
Whenever I watch a movie with my friends and one of them has to transcode, that friend tend to be about 2 seconds behind. Pausing and resuming will occasionally result in further desync. We've been testing this with both Jellyfin Media Player and in a browser. When all of us have to transcode, we get even more severe issues with this.

**Steps To Reproduce**
1. Start a movie/show
2. Enable syncplay between two or more users
3. Assuming at least one of the users is transcoding, this should be happening

**Expected Behavior**
We should stay in sync regardless of whether we're transcoding or not

**Logs**
I can try to provide if necessary. Don't have recent logs to look at and the logs can be annoying to parse, so I'll wait for this to be requested.

**System (please complete the following information):**
 - Platform: Linux/Flatpak, MacOS, Windows (assumedly), as long as it's transcoding
 - Browser: Firefox, Chrome, Safari as long as it's transcoding
 - Jellyfin Version: 10.9.0

**Additional Context**
I'm wondering if this could be related to this issue #4346 as the syncplay desync seems to be similarly long to the subtitle desync.

**Comment by thornbill** [comment]
Have you tried enabling/changing the sync correction settings in the syncplay menu while in a session?

**Comment by Reinachan** [comment]
I tried it by running the same movie in Chrome (because I couldn't enable sync correction in Firefox) and in Jellyfin Media Player with sync correction on and the movies are not in sync. Both were transcoding as it was a DolbyVision movie.

Then I tried it using a movie that only had to transcode in Chrome due to unsupported container format (mkv). This did stay in sync when sync correction was enabled. When I disabled it and seeked around, the video got desynced. When I re-enabled it, it synced again.

It's great to know that there is a way to get it working as long as only one of us has to transcode, but what about if I want to watch a movie in Dolby Vision on my TV while my friends transcodes it? Or in case that would work as there is a device that isn't transcoding; what if I don't want to bother with the hassle of streaming it to my TV and just want to watch it on my computer so I can watch it with my friends without bothering them with the movie being on speaker? In that case, this wouldn't really work.

My point is that I would still like the issue to be properly resolved even if there seems to be a workaround that works in some situations :)

**Comment by thornbill** [comment]
Yes, I fully expected it would not be fully fixed but at least it can be _somewhat_ worked around. Unfortunately until someone is willing to essentially rewrite how syncplay works this is unlikely to be fixed.

**Comment by bluemoehre** [comment]
I face that issue in WebOS. Depending on the bandwidth of audio the desync is up to 1sec (Dolby Atmos) =/

**Comment by Reinachan** [comment]
Yeah, I had that issue on WebOS too. It seemed like the audio desync was proportional to the desync I had from my friend when syncplaying with them

**Comment by Krupski19** [comment]
I too have been experiencing this issue, and have experimented with a few things, made sure that Sync Corrections were enabled on all clients and so on, and at least for me, the desync issue is caused by having Prefer fMP4-HLS Media Container enabled in the client's settings under Settings -> Playback.

When a  user transcodes a video with this setting enabled, it seems as though the timestamp of the playback of the video matches that of the client that is direct streaming the video and with what the server is reporting, but the frame and audio displayed on the transcoded video with Prefer fMP4-HLS Media Container enabled seems to be behind by a few seconds.

Disabling Prefer fMP4-HLS Media Container on the clients setting and watching the same video again in sync play and it being transcoded now has the video synced perfectly with the devices direct streaming the content in the sync play group.  However disabling Prefer fMP4-HLS Media Container prevents the video from being transcoded to HEVC or AV1, and can only be transcoded to H264.

As requiring to remind my users every time they login to a new client or browser, or when their cookies get deleted, to go in and enable Sync Corrections and disable Prefer fMP4-HLS Media Container to have proper sync play playback is a pain. Could we instead implement a feature to allow the server or server administrator to push default client settings for all clients that connect to it?  These settings could later be overridden by the user if they wish to do so, but upon first sign in, settings such as Sync Corrections and Prefer fMP4-HLS Media Container would already be set correctly to resolve this and other issues clients may face.

As reading on the complexity of resolving this syncing issue with fMP4-HLS containers is quite difficult, I believe this would be a good compromise to resolve syncplay playback issues for the mean time while.

**Comment by sambartik** [comment]
> Yes, I fully expected it would not be fully fixed but at least it can be _somewhat_ worked around. Unfortunately until someone is willing to essentially rewrite how syncplay works this is unlikely to be fixed.

Could you please elaborate on that? Do you have something specific in mind with the rewrite?

**Comment by PeachesMLG** [comment]
Out of curiosity if you look at the timestamps on the video do they match?

I raised a bug report here https://github.com/jellyfin/jellyfin/issues/13965 with some info on this here, and when you transcode starting from a specific place, it wont start at the exact location which causes a lot of the desync issues caused by the -noaccurate_seek flag.

You can even replicate this using raw ffmpeg by setting a `-ss 00:14:54.000` (e.g. start at 14 minutes) and adding -noaccurate_seek

When I did this ffmpeg actually started the video playback at 14:46 not 14:54 (even more annoyingly later on ffmpeg just decides to skip the video forwards about 10 seconds completly cutting stuff out)

So tldr; the web video player thinks it is on 14:54 but is actually on 14:46.
If this is the case with your issue then its most likely nothing todo with syncplay and is instead todo with this ffmpeg issue

**Comment by Mumbolio85** [comment]
I thought I would take a look at this issue, and even ended up rewriting the whole SyncPlay functionality in the process. What I discovered is that even with my rewritten code, which feels less finicky than the existing code (I would be willing to open a PR with my revised SyncPlay code even though it does not solve this issue), the video on one of the devices (assuming at least one is transcoding) will be between 2-5 seconds behind the other. 

As I continued to work out what the problem was, I discovered that the timestamps **_WERE_** perfectly in time with each other. I remembered the mention of -noaccurate_seek above and looked into that. I fully agree with the comment above that -noaccurate_seek is the culprit as the code uses the timestamps to sync them, and since the timestamp of the transcoding device is 2-5 seconds off (even at the very start of the movie), there is not much that rewriting the SyncPlay code can do to help this issue. 

From what I can tell after having followed the chain of issues referencing -noaccurate_seek, this issue being solved relies upon people coming to a decision regarding what to do with the feature, and that brings its own set of discussions to the table. However, I am more than willing to tidy up my revised SyncPlay code and open a PR for it even though it does not solve this issue

**Comment by robert-w-gries** [comment]
I've been down a rabbit hole debugging SyncPlay de-sync issues the past couple weeks, and I noticed this frontend PR that got merged a couple days ago, #7473, with the related backend PR, jellyfin/jellyfin#16580

It seems this updated default seek strategy will replace the `-noaccurate_seek` flag with a similarly cheap but now **accurate** seek flag. This new seek strategy is expected to be released in Jellyfin 12.0. Please correct me if I'm mistaken.

Edit: I tested v12-rc5 with a group of friends, and it was a massive improvement. There was only a couple times where someone de-synced, and it was quickly fixed by pausing and then resuming.

Prior to v12, it was almost guaranteed that someone in the group would be off by 5-15 seconds.
